src/libraries/Native/Unix/System.Native/pal_networking.c

```diff
@@ -432,12 +421,148 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
     return ret;
 }
 
+#if HAVE_GETADDRINFO_A
+struct GetAddrInfoAsyncState
+{
+    struct gaicb gai_request;
+    struct gaicb* gai_requests;
+    struct sigevent sigevent;
+
+    HostEntry* entry;
+    GetHostEntryForNameCallback callback;
+    char address[];
+};
+
+static void GetHostEntryForNameAsyncComplete(sigval_t context)
+{
+    struct GetAddrInfoAsyncState* state = (struct GetAddrInfoAsyncState*)context.sival_ptr;
+
+    atomic_thread_fence(memory_order_acquire);
+
+    GetHostEntryForNameCallback callback = state->callback;
+
+    int ret = ConvertGetAddrInfoAndGetNameInfoErrorsToPal(gai_error(&state->gai_request));
+
+    if (ret == 0)
+    {
+        const uint8_t* address = (const uint8_t*)state->address;
+        struct addrinfo* info = state->gai_request.ar_result;
+
+        ret = GetHostEntries(address, info, state->entry);
+    }
+
+    if (callback != NULL)
+    {
+        callback(state->entry, ret);
+    }
+
+    free(state);
+}
+#endif
+
+int32_t SystemNative_PlatformSupportsGetAddrInfoAsync()
+{
+    return HAVE_GETADDRINFO_A;
+}
+
+// Get all address families and the canonical name
+static const struct addrinfo s_hostEntryForNameHints = {.ai_family = AF_UNSPEC, .ai_flags = AI_CANONNAME};
+
+int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entry)
+{
+    if (address == NULL || entry == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    struct addrinfo* info = NULL;
+
+    int result = getaddrinfo((const char*)address, NULL, &s_hostEntryForNameHints, &info);
+    if (result != 0)
+    {
+        return ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+    }
+
+    return GetHostEntries(address, info, entry);
+}
+
+int32_t SystemNative_GetHostEntryForNameAsync(const uint8_t* address, HostEntry* entry, GetHostEntryForNameCallback callback)
+{
+#if HAVE_GETADDRINFO_A
+    if (address == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    size_t addrlen = strlen((const char*)address);
+
+    if (addrlen > _POSIX_HOST_NAME_MAX)
```

Comment:
`_POSIX_HOST_NAME_MAX` is fixed to 255. It's more than the 253 according the RFC (cf. https://github.com/dotnet/runtime/pull/34633#discussion_r406447963), but it matches 
https://github.com/dotnet/runtime/blob/458579b71af2456bc97e410ecfb99219309f5cf5/src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetHostName.cs#L18

Reply:
BTW: neither the Windows-Pal nor the sync-version have a check for max hostname. Should these be added (at the managed side to catch them early)? Or leave it as the uncommon case and let the OS handle it?

Reply:
In this case, it's better to check because you're computing the allocation length based on  what is potentially user input; that could overflow and become an issue. By capping it to `POSIX_HOST_NAME_MAX` like you did here, the overflow is avoided (and there's no need to perform an overflow check).

In other cases, it's fine to leave it to the OS to verify, as they're passing the string as it is to the system calls/library functions.